File: src/services/tauri/docker.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  DockerContainer,
  DockerContainerAction,
  DockerContainerDetails,
  DockerImage,
  DockerImageAction,
  DockerNetwork,
  DockerNetworkAction,
  DockerOverview,
  DockerSystemAction,
  DockerVolume,
  DockerVolumeAction,
} from "@/types/docker";

export function checkRemoteDocker(): Promise<boolean> {
  return invoke<boolean>("check_remote_docker");
}

export function listRemoteDockerContainers(): Promise<DockerContainer[]> {
  return invoke<DockerContainer[]>("list_remote_docker_containers");
}

export function listRemoteDockerImages(): Promise<DockerImage[]> {
  return invoke<DockerImage[]>("list_remote_docker_images");
}

export function listRemoteDockerNetworks(): Promise<DockerNetwork[]> {
  return invoke<DockerNetwork[]>("list_remote_docker_networks");
}

export function listRemoteDockerVolumes(): Promise<DockerVolume[]> {
  return invoke<DockerVolume[]>("list_remote_docker_volumes");
}

export function runRemoteDockerContainerAction(
  containerId: string,
  action: DockerContainerAction,
): Promise<string> {
  return invoke<string>("run_remote_docker_container_action", {
    containerId,
    action,
  });
}

export function runRemoteDockerImageAction(
  imageRef: string,
  action: DockerImageAction,
): Promise<string> {
  return invoke<string>("run_remote_docker_image_action", {
    imageRef,
    action,
  });
}

export function runRemoteDockerVolumeAction(
  volumeName: string,
  action: DockerVolumeAction,
): Promise<string> {
  return invoke<string>("run_remote_docker_volume_action", {
    volumeName,
    action,
  });
}

export function runRemoteDockerNetworkAction(
  networkRef: string,
  action: DockerNetworkAction,
): Promise<string> {
  return invoke<string>("run_remote_docker_network_action", {
    networkRef,
    action,
  });
}

export function runRemoteDockerSystemAction(
  action: DockerSystemAction,
): Promise<string> {
  return invoke<string>("run_remote_docker_system_action", {
    action,
  });
}

export function getRemoteDockerContainerDetails(
  containerId: string,
): Promise<DockerContainerDetails> {
  return invoke<DockerContainerDetails>("get_remote_docker_container_details", {
    containerId,
  });
}

export function getRemoteDockerContainerLogs(
  containerId: string,
  options?: { tail?: number; since?: string },
): Promise<string> {
  return invoke<string>("get_remote_docker_container_logs", {
    containerId,
    tail: options?.tail ?? 200,
    since: options?.since?.trim() ? options.since.trim() : null,
  });
}

export function getRemoteDockerOverview(): Promise<DockerOverview> {
  return invoke<DockerOverview>("get_remote_docker_overview");
}

File: src-tauri/src/commands/docker.rs
use tauri::State;

use crate::services::DockerContainer;
use crate::services::DockerContainerDetails;
use crate::services::DockerImage;
use crate::services::DockerNetwork;
use crate::services::DockerOverview;
use crate::services::DockerVolume;
use crate::services::SshConnectionManager;

// `async` 속성: sync 본문이라도 메인 스레드가 아닌 blocking pool에서 실행되어
// SSH I/O 중 웹뷰 페인트(모달 즉시 표시)가 막히지 않는다.
#[tauri::command(async)]
pub fn check_remote_docker(state: State<'_, SshConnectionManager>) -> Result<bool, String> {
    state.is_docker_installed()
}

#[tauri::command(async)]
pub fn list_remote_docker_containers(
    state: State<'_, SshConnectionManager>,
) -> Result<Vec<DockerContainer>, String> {
    state.list_docker_containers()
}

#[tauri::command(async)]
pub fn list_remote_docker_images(
    state: State<'_, SshConnectionManager>,
) -> Result<Vec<DockerImage>, String> {
    state.list_docker_images()
}

#[tauri::command(async)]
pub fn list_remote_docker_networks(
    state: State<'_, SshConnectionManager>,
) -> Result<Vec<DockerNetwork>, String> {
    state.list_docker_networks()
}

#[tauri::command(async)]
pub fn list_remote_docker_volumes(
    state: State<'_, SshConnectionManager>,
) -> Result<Vec<DockerVolume>, String> {
    state.list_docker_volumes()
}

#[tauri::command(async)]
pub fn get_remote_docker_overview(
    state: State<'_, SshConnectionManager>,
) -> Result<DockerOverview, String> {
    state.get_docker_overview()
}

#[tauri::command(async)]
pub fn run_remote_docker_container_action(
    container_id: String,
    action: String,
    state: State<'_, SshConnectionManager>,
) -> Result<String, String> {
    state.docker_container_action(&container_id, &action)
}

#[tauri::command(async)]
pub fn run_remote_docker_image_action(
    image_ref: String,
    action: String,
    state: State<'_, SshConnectionManager>,
) -> Result<String, String> {
    state.docker_image_action(&image_ref, &action)
}

#[tauri::command(async)]
pub fn run_remote_docker_volume_action(
    volume_name: String,
    action: String,
    state: State<'_, SshConnectionManager>,
) -> Result<String, String> {
    state.docker_volume_action(&volume_name, &action)
}

#[tauri::command(async)]
pub fn run_remote_docker_network_action(
    network_ref: String,
    action: String,
    state: State<'_, SshConnectionManager>,
) -> Result<String, String> {
    state.docker_network_action(&network_ref, &action)
}

#[tauri::command(async)]
pub fn run_remote_docker_system_action(
    action: String,
    state: State<'_, SshConnectionManager>,
) -> Result<String, String> {
    state.docker_system_action(&action)
}

#[tauri::command(async)]
pub fn get_remote_docker_container_details(
    container_id: String,
    state: State<'_, SshConnectionManager>,
) -> Result<DockerContainerDetails, String> {
    state.docker_container_details(&container_id)
}

#[tauri::command(async)]
pub fn get_remote_docker_container_logs(
    container_id: String,
    tail: u32,
    since: Option<String>,
    state: State<'_, SshConnectionManager>,
) -> Result<String, String> {
    state.docker_container_logs(&container_id, tail, since.as_deref().unwrap_or(""))
}
